util/extsort/disk_sorter.go

```diff
@@ -131,29 +900,363 @@ func (d *DiskSorter) Sort(ctx context.Context) error {
+	}
+	return nil
+}
+
+func (d *DiskSorter) compactFiles(ctx context.Context, files []*fileMetadata) error {
+	// Split files into multiple compaction groups.
+	// Each group will be compacted independently.
+	groups := splitCompactionFiles(files, d.opts.MaxCompactionDepth)
+	compactions := make([]*compaction, 0, len(groups))
+	for _, group := range groups {
+		compactions = append(compactions, buildCompactions(group, d.opts.MaxCompactionSize)...)
+	}
+
+	// Build file references.
+	fileRefs := make(map[int]*atomic.Int32)
+	for _, file := range files {
+		fileRefs[file.fileNum] = new(atomic.Int32)
+	}
+	for _, c := range compactions {
+		for _, file := range c.overlapFiles {
+			fileRefs[file.fileNum].Add(1)
+		}
+	}
+	d.opts.Logger.Info("total compactions", zap.Int("count", len(compactions)))
+
+	// Run all compactions.
+	removedFileNums := generic.NewSyncMap[int, struct{}](len(files))
+	outputCh := make(chan *fileMetadata, len(compactions))
+	g, gCtx := errgroup.WithContext(ctx)
+	g.SetLimit(d.opts.Concurrency)
+	for _, c := range compactions {
+		c := c
+		g.Go(func() error {
+			if gCtx.Err() != nil {
+				return errors.Trace(gCtx.Err())
+			}
+			output, err := d.runCompaction(gCtx, c)
+			if err != nil {
+				return err
+			}
+			outputCh <- output
+			for _, file := range c.overlapFiles {
+				if fileRefs[file.fileNum].Add(-1) == 0 {
+					_ = d.fs.Remove(makeFilename(d.fs, d.dirname, file.fileNum))
+					removedFileNums.Store(file.fileNum, struct{}{})
+				}
+			}
+			return nil
+		})
+	}
+	if err := g.Wait(); err != nil {
+		return err
+	}
+
+	// Update d.orderedFiles.
+	close(outputCh)
+	for output := range outputCh {
+		d.orderedFiles = append(d.orderedFiles, output)
+	}
+	var newOrderedFiles []*fileMetadata
+	for _, file := range d.orderedFiles {
+		if _, ok := removedFileNums.Load(file.fileNum); !ok {
+			newOrderedFiles = append(newOrderedFiles, file)
+		}
+	}
+	slices.SortFunc(newOrderedFiles, func(a, b *fileMetadata) bool {
+		return bytes.Compare(a.startKey, b.startKey) < 0
+	})
+	d.orderedFiles = newOrderedFiles
 	return nil
 }
 
-func (d *DiskSorter) doSort(_ context.Context) error {
-	iter := d.db.NewIter(nil)
-	if !iter.Last() {
-		// No keys or any error occurred.
+func pickCompactionFiles(
+	allFiles []*fileMetadata,
+	compactionThreshold int,
+	logger *zap.Logger,
+) []*fileMetadata {
+	type interval struct {
+		key   []byte
+		depth int
+	}
+	// intervals is a list of intervals that perfectly overlap file boundaries.
+	// For example, if we have two files [a, c] and [b, d], the intervals will be
+	// [a, b), [b, c), [c, d).
+	intervals := make([]interval, 0, len(allFiles)*2)
+	for _, file := range allFiles {
+		intervals = append(intervals, interval{
+			key:   file.startKey,
+			depth: 1,
+		})
+		intervals = append(intervals, interval{
+			key:   file.endKey,
+			depth: -1,
+		})
+	}
+	slices.SortFunc(intervals, func(a, b interval) bool {
+		return bytes.Compare(a.key, b.key) < 0
```

Comment:
can we also place depth=-1 before depth=1 when keys are equal? so for [a, b) and [b, c) we will not get maxDepth=2

Reply:
Good catch! Addressed in https://github.com/pingcap/tidb/pull/44185/commits/c5b76a8e3371aaefeaaed1f11a428a3e1258eb7e.